Hw9 jpa

demen777 opened the pull request:
Переписать приложение для хранения книг на ORM
Цель: Цель: полноценно работать с JPA + Hibernate для подключения к реляционным БД посредством ORM-фреймворка Результат: Высокоуровневое приложение с JPA-маппингом сущностей
Домашнее задание выполняется переписыванием предыдущего на JPA.

Требования:
1. Использовать JPA, Hibernate только в качестве JPA-провайдера.
2. Добавить комментарии к книгам, и высокоуровневые сервисы, оставляющие комментарии к книгам.
3. Покрыть репозитории тестами, используя H2 базу данных и соответствующий H2 Hibernate-диалект для тестов.
4. Не забудьте отключить DDL через Hibernate

demen777 committed:
Migrate to JPA
Author tests green

demen777 committed:
Genre tests green

demen777 committed:
Book tests green

demen777 committed:
Add clear context to AuthorDaoJpaTest

demen777 committed:
First full implementation hw9_jpa

demen777 committed:
Move book-comments link to Book model

demen777 committed:
Add ServiceTestConfiguration for service tests context caching

demen777 committed:
Change TestConfiguration to Configuration for ServiceTestConfiguration and add ComponentScan in services package

demen777 committed:
Use ContextConfiguration instead SpringBootTest in services tests

demen777 committed:
Add liquibase support

demen777 committed:
Use lambda in shell components for handling ServiceError exception
Fix error application.yml

demen777 committed:
Remove not used BookCommentService.findByBookIdMethod

stvort commented on hw9_jpa/src/main/java/ru/otus/demen/books/model/Book.java:
```diff
@@ -0,0 +1,51 @@
+
+    @OneToMany(fetch = FetchType.EAGER, cascade = CascadeType.ALL)
+    @JoinColumn(name = "book_id", nullable = false)
+    private List<BookComment> bookComments = new ArrayList<>();
```
1. Не уверен, что комментарии это свойство книги, скорее наоборот, информация о книге это свойство комментария
2. Так или иначе доставать книги сразу с комментариями это явно избыточно. Будет лучше получать их только тогда, когда они действительно требуются

demen777 replied:
bookComments удалено из Book
В BookComment добавлен book

stvort commented on hw9_jpa/src/main/java/ru/otus/demen/books/model/Book.java:
```diff
@@ -0,0 +1,51 @@
+    @Override
+    public int hashCode() {
+        return Objects.hash(name, author);
+    }
```
Т.е. две книги с одним названием и автором, но разными жанрами считаем одинаковыми?

demen777 replied:
Добавил поле genre в equals и hashCode

stvort commented on hw9_jpa/src/main/java/ru/otus/demen/books/dao/BookCommentDaoJpa.java:
```diff
@@ -0,0 +1,31 @@
+    @Override
+    public Collection<BookComment> findByBookId(long bookId) {
+        TypedQuery<BookComment> query = em.createQuery(
+                "select bc from BookComment bc join bc.book b where b.id = :bookId",
```
Предлагаю включить ```spring.jpa.show-sql=true``` и посмотреть к какому количеству запросов SQL приведет выполнение данного запроса. Лучше воспользоваться join fetch или EntityGraph для авторов и жанров, плюс что-то подобрать для комментариев если они останутся в книге

demen777 replied:
Использовал join fetch

demen777 replied:
select
        bookcommen0_.id as id1_2_0_,
        book1_.id as id1_1_1_,
        author2_.id as id1_0_2_,
        genre3_.id as id1_3_3_,
        bookcommen0_.book_id as book_id3_2_0_,
        bookcommen0_.text as text2_2_0_,
        book1_.author_id as author_i3_1_1_,
        book1_.genre_id as genre_id4_1_1_,
        book1_.name as name2_1_1_,
        author2_.first_name as first_na2_0_2_,
        author2_.surname as surname3_0_2_,
        genre3_.name as name2_3_3_ 
    from
        book_comments bookcommen0_ 
    inner join
        books book1_ 
            on bookcommen0_.book_id=book1_.id 
    inner join
        authors author2_ 
            on book1_.author_id=author2_.id 
    inner join
        genres genre3_ 
            on book1_.genre_id=genre3_.id 
    where
        book1_.id=?

stvort commented on hw9_jpa/src/main/resources/data.sql:
```diff
@@ -0,0 +1,3 @@
+insert into authors(id, first_name, surname) values (1, 'Лев', 'Толстой');
```
Данные тоже можно добавить через liquibase. В примере, в репозитории это есть

demen777 replied:
Добавил инициализацию данных в liquibase

stvort commented on hw9_jpa/src/test/java/ru/otus/demen/books/dao/AuthorDaoJpaTest.java:
```diff
@@ -0,0 +1,92 @@
+    @DisplayName("Поиск методом findById не нашел автора")
+    void findById_authorNotFound() {
+        Optional<Author> author = authorDao.findById(WRONG_AUTHOR_ID);
+        assertThat(author.isPresent()).isFalse();
```
`assertThat(author).isPresent();`

demen777 replied:
Изменил на assertThat(author).isEmpty();

stvort commented on hw9_jpa/src/test/java/ru/otus/demen/books/service/AuthorServiceImplTest.java:
```diff
@@ -0,0 +1,145 @@
+import static org.mockito.Mockito.when;
+
+@ExtendWith(SpringExtension.class)
+@ContextConfiguration(classes = ServiceTestConfiguration.class)
```
Давайте попробуем сделать то же самое, но со `@SpringBootTest`

demen777 replied:
Изменил на @SpringBootTest

Commit:
Use SpringBootTest in services tests
Use liquibase for data initialization
Use assertThat(genre).isEmpty() instead assertThat(genre.isPresent()).isFalse() in GenreServiceImplTest

Commit:
Add genre field to equals and hashCode

Commit:
Remove bookComments from Book and add book to BookComment

Commit:
Use isPresent and isEmpty in test asserts

Commit:
Add join fetch to Collection<BookComment> findByBookId(long bookId)
